src/server/qtquick/wtextureproviderprovider.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2024 UnionTech Software Technology Co., Ltd.
+// SPDX-License-Identifier: Apache-2.0 OR LGPL-3.0-only OR GPL-2.0-only OR GPL-3.0-only
+
+#include "wtextureproviderprovider.h"
+#include "rhi/qrhi.h"
+#include "woutputrenderwindow.h"
+#include "private/qquickwindow_p.h"
+#include "wthreadutils.h"
+
+extern "C" {
+#include <wlr/types/wlr_buffer.h>
+#include <wlr/render/wlr_texture.h>
+#include <wlr/render/wlr_renderer.h>
+}
+WAYLIB_SERVER_BEGIN_NAMESPACE
+Q_LOGGING_CATEGORY(qLcTextureProvider, "waylib.server.texture.provider")
+
+WTextureProviderProvider::WTextureProviderProvider() { }
+
+static void cleanupRbResult(void *rbResult)
+{
+    delete reinterpret_cast<QRhiReadbackResult*>(rbResult);
+}
+
+class RenderToImageJob : public QObject, public QRunnable
+{
+    Q_OBJECT
+public:
+    RenderToImageJob(WOutputRenderWindow *renderWindow, WTextureProviderProvider *provider, QPromise<QImage> promise)
+        : m_renderWindow(renderWindow)
+        , m_provider(provider)
+        , m_promise(std::move(promise))
+    { }
+
+    void run() override {
+        if (m_promise.isCanceled())
+            return;
+        m_promise.start();
+        WSGTextureProvider *textureProvider = m_provider->wTextureProvider();
+        if (textureProvider && textureProvider->texture() && textureProvider->texture()->rhiTexture()) {
+            // Perform rhi texture read back
+            auto texture = textureProvider->texture()->rhiTexture();
+            qCInfo(qLcTextureProvider) << "Perform rhi texture read back for texture" << texture;
+            QRhiReadbackResult *rbResult = new QRhiReadbackResult;
+            QRhiCommandBuffer *cb;
+            m_renderWindow->rhi()->beginOffscreenFrame(&cb);
+            auto ub = m_renderWindow->rhi()->nextResourceUpdateBatch();
+            cb->beginComputePass(ub);
+            QRhiReadbackDescription rbd(texture);
+            ub->readBackTexture(rbd, rbResult);
+            cb->endComputePass(ub);
+            auto frameOpResult = m_renderWindow->rhi()->endOffscreenFrame();
+            if (frameOpResult == QRhi::FrameOpSuccess) {
+                m_promise.addResult(QImage(reinterpret_cast<const uchar *>(rbResult->data.constData()),
+                                           rbResult->pixelSize.width(),
+                                           rbResult->pixelSize.height(),
+                                           QImage::Format_RGBA8888_Premultiplied,
```

Comment:
Why is Format_RGBA8888_Premultiplied

Reply:
Just a workaround here now, we can't acquire the accurate format unless we upgrade to 0.18.

Reply:
Add a todo

Reply:
Already added at next line, the end of the invokation.